File: Data Structures & Algorithms/duplicate-integer/submission-4.java
class Solution {
    public boolean hasDuplicate(int[] nums) {
        /**
        Time: O(n^2)
        Space: O(1)
        */

        for (int index = 0; index < nums.length; index++) {
            for (int j = index + 1; j < nums.length; j++) {
                if (nums[index] == nums[j]) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: Data Structures & Algorithms/is-palindrome/submission-4.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        """
        Check the string in place using 2 pointers: l, r
        - l starts at index 0: traverse forward
        - r starts at index n - 1: traverse backward
        Time: O(n)
        Space: O(1)
        """

        l = 0
        r = len(s) - 1
        while l < r:
            while l < r and not s[l].isalnum():
                l += 1
            while l < r and not s[r].isalnum():
                r -= 1
            if s[l].lower() != s[r].lower():
                return False
            l += 1
            r -= 1
                
        return True